File: sdk/node/src/proxy.ts
// Zero dependencies — works in Node.js, Deno, Bun, Edge runtimes.

import crypto from 'crypto'

const DEFAULT_PROXY =
  process.env.AGENTGUARD_URL ?? 'http://localhost:7777'

/**
 * Wrap an upstream URL with AgentGuard enforcement.
 *
 * Returns the AgentGuard proxy URL.
 * Pass this to any MCP or HTTP client instead of the real URL.
 *
 * @example
 * // Before
 * const client = new MCPClient('http://my-mcp-server:8080')
 *
 * // After (one word change)
 * const client = new MCPClient(guard('http://my-mcp-server:8080'))
 */
export function guard(
  upstreamUrl: string,
  options?: {
    proxyUrl?: string
    agentId?:  string
  }
): string {
  const base = (options?.proxyUrl ?? DEFAULT_PROXY).replace(/\/$/, '')
  const encoded = encodeURIComponent(upstreamUrl)
  let url = `${base}?upstream=${encoded}`
  if (options?.agentId) {
    url += `&agent_id=${encodeURIComponent(options.agentId)}`
  }
  return url
}

export interface EscalationItem {
  id:          string
  agent_id:    string
  tool_name:   string
  protocol:    string
  arguments:   Record<string, unknown>
  ts:          string
  expires_at:  string
  age_minutes: number
}

export interface HealthStatus {
  status:      string
  version:     string
  uptime_s:    number
  environment: string
  goroutines:  number
}

/**
 * AgentGuard client with full API access.
 *
 * @example
 * const ag = new AgentGuard({
 *   proxyUrl:  'http://agentguard:7777',
 *   agentId:   'billing-agent',
 *   adminKey:  process.env.AGENTGUARD_ADMIN_KEY,
 * })
 *
 * const client = new MCPClient(ag.wrap('http://my-api:8080'))
 * const pending = await ag.listEscalations()
 * await ag.approve(pending.id)
 */
export class AgentGuard {
  private proxyUrl: string
  private agentId?: string
  private signingSecret: string
  private adminKey: string

  constructor(options?: {
    proxyUrl?: string
    agentId?:  string
    signingSecret?: string
    adminKey?: string
  }) {
    this.proxyUrl = (options?.proxyUrl ?? DEFAULT_PROXY).replace(/\/$/, '')
    this.agentId  = options?.agentId
    this.signingSecret = options?.signingSecret ?? process.env.AGENTGUARD_SIGNING_SECRET ?? ''
    this.adminKey = options?.adminKey
      ?? process.env.AGENTGUARD_ADMIN_KEY
      ?? ''
  }

  /**
   * Build HMAC-SHA256 signature over canonical string.
   * Returns "sha256=<hex>"
   */
  private signRequest(body: string, timestamp: string): string {
    const bodyHash = crypto
      .createHash('sha256')
      .update(body)
      .digest('hex')

    const canonical =
      `${this.agentId}\n${timestamp}\n${bodyHash}`

    const sig = crypto
      .createHmac('sha256', this.signingSecret)
      .update(canonical)
      .digest('hex')

    return `sha256=${sig}`
  }

  /**
   * Returns auth headers for a request body.
   * Returns empty object if no signing secret (dev mode).
   */
  signedHeaders(body: string): Record<string, string> {
    if (!this.signingSecret || !this.agentId) return {}

    const ts  = String(Math.floor(Date.now() / 1000))
    const sig = this.signRequest(body, ts)

    return {
      'X-Agent-ID':             this.agentId,
      'X-AgentGuard-Timestamp': ts,
      'X-AgentGuard-Signature': sig,
    }
  }

  /**
   * Make a signed fetch call through AgentGuard.
   * Automatically adds auth headers.
   *
   * Usage:
   *   const result = await guard.fetch({
   *     method:  'tools/call',
   *     params:  { name: 'read_file', arguments: {} },
   *   })
   */
  async fetch(payload: {
    method:  string
    params?: Record<string, unknown>
    id?:     string | number
  }): Promise<Response> {
    const body = JSON.stringify({
      jsonrpc: '2.0',
      id:      payload.id ?? crypto.randomUUID(),
      method:  payload.method,
      params:  payload.params ?? {},
    })

    return globalThis.fetch(this.proxyUrl, {
      method:  'POST',
      headers: {
        'Content-Type': 'application/json',
        ...this.signedHeaders(body),
      },
      body,
    })
  }

  /** Wrap an upstream URL with this instance's config. */
  wrap(upstreamUrl: string): string {
    return guard(upstreamUrl, {
      proxyUrl: this.proxyUrl,
      agentId:  this.agentId,
    })
  }

  /** List pending escalations. */
  async listEscalations(): Promise<EscalationItem[]> {
    try {
      const res = await fetch(`${this.proxyUrl}/escalations`, {
        headers: { 'X-AgentGuard-Admin-Key': this.adminKey },
      })
      const data = await res.json() as { escalations: EscalationItem[] }
      return data.escalations ?? []
    } catch {
      return []
    }
  }

  /** Approve a pending escalation. */
  async approve(escalationId: string): Promise<boolean> {
    return this._resolve(escalationId, 'approve')
  }

  /** Reject a pending escalation. */
  async reject(escalationId: string): Promise<boolean> {
    return this._resolve(escalationId, 'reject')
  }

  /** Check AgentGuard health. */
  async health(): Promise<HealthStatus | null> {
    try {
      const res = await fetch(`${this.proxyUrl}/health`)
      return await res.json() as HealthStatus
    } catch {
      return null
    }
  }

  private async _resolve(
    id:     string,
    action: 'approve' | 'reject',
  ): Promise<boolean> {
    try {
      const res = await fetch(
        `${this.proxyUrl}/escalations/${id}/${action}`,
        {
          method:  'POST',
          headers: { 'X-AgentGuard-Admin-Key': this.adminKey },
        }
      )
      return res.ok
    } catch {
      return false
    }
  }
}

File: sdk/python/agentguard/proxy.py
# agentguard/proxy.py
# Core AgentGuard proxy wrapper.
# guard() is a one-line drop-in replacement.
# Instead of pointing your MCP/HTTP client at your real API,
# you point it at AgentGuard which enforces your policy first.
# How it works:
#   guard("http://real-api:8080")
#   → returns "http://localhost:7777"
#   → sets X-AgentGuard-Upstream header so AgentGuard knows
#     where to forward allowed requests
# The real API URL is passed as a header, not in the URL.
# This means your code doesn't change at all — just swap the URL.

import os
import urllib.parse
import hashlib
import hmac as hmac_lib
import time
from typing import Optional


# Default AgentGuard proxy URL
_DEFAULT_PROXY = os.environ.get("AGENTGUARD_URL", "http://localhost:7777")


def guard(
    upstream_url: str,
    proxy_url:    Optional[str]  = None,
    agent_id:     Optional[str]  = None,
) -> str:
    """
    Wrap an upstream URL with AgentGuard enforcement.

    Returns the AgentGuard proxy URL. Pass this to any
    MCP or HTTP client instead of the real upstream URL.

    Args:
        upstream_url: The real API URL you want to call.
        proxy_url:    AgentGuard URL (default: AGENTGUARD_URL env
                      or http://localhost:7777)
        agent_id:     Optional agent identifier for policy matching.

    Returns:
        str: The AgentGuard proxy URL to use instead.

    Example:
        from agentguard import guard
        from mcp import MCPClient

        # Before:
        client = MCPClient("http://my-mcp-server:8080")

        # After (one word change):
        client = MCPClient(guard("http://my-mcp-server:8080"))
    """
    base = (proxy_url or _DEFAULT_PROXY).rstrip("/")

    # Encode upstream as query param so the proxy knows
    # where to forward (alternative: use X-Upstream header)
    encoded = urllib.parse.quote(upstream_url, safe="")
    url     = f"{base}?upstream={encoded}"

    if agent_id:
        url += f"&agent_id={urllib.parse.quote(agent_id)}"

    return url


class AgentGuard:
    """
    AgentGuard client with configuration.
    Use this when you need more control than guard() provides.

    Example:
        ag = AgentGuard(
            proxy_url="http://agentguard:7777",
            agent_id="billing-agent",
            admin_key="secret",
        )

        # Wrap a URL
        client = MCPClient(ag.wrap("http://my-api:8080"))

        # List pending escalations
        escalations = ag.list_escalations()

        # Approve an escalation
        ag.approve("escalation-id")
    """

    def __init__(
        self,
        proxy_url:       Optional[str] = None,
        agent_id:        Optional[str] = None,
        signing_secret:  Optional[str] = None,
        admin_key:       Optional[str] = None,
    ):
        self.proxy_url = (proxy_url or _DEFAULT_PROXY).rstrip("/")
        self.agent_id  = agent_id
        self.signing_secret = signing_secret or os.environ.get("AGENTGUARD_SIGNING_SECRET", "")
        self.admin_key = admin_key or os.environ.get("AGENTGUARD_ADMIN_KEY", "")

    def _sign_request(
        self,
        body:      bytes,
        timestamp: str,
    ) -> str:
        """
        Build HMAC-SHA256 signature over canonical string.
        Returns "sha256=<hex>"
        """
        body_hash = hashlib.sha256(body).hexdigest()
        canonical = f"{self.agent_id}\n{timestamp}\n{body_hash}"
        sig = hmac_lib.new(
            self.signing_secret.encode(),
            canonical.encode(),
            hashlib.sha256,
        ).hexdigest()
        return f"sha256={sig}"

    def signed_headers(self, body: bytes) -> dict:
        """
        Returns the auth headers to add to every request.
        If no signing_secret configured, returns empty dict.
        (backward compatible — dev mode still works)
        """
        if not self.signing_secret or not self.agent_id:
            return {}

        ts = str(int(time.time()))
        sig = self._sign_request(body, ts)
        return {
            "X-Agent-ID":              self.agent_id,
            "X-AgentGuard-Timestamp":  ts,
            "X-AgentGuard-Signature":  sig,
        }

    def wrap_request(
        self,
        body:    bytes,
        headers: dict = None,
    ) -> dict:
        """
        Returns a complete headers dict ready to attach to
        any HTTP request to the AgentGuard proxy.
        """
        base = {
            "Content-Type": "application/json",
            "X-Agent-ID":   self.agent_id,
        }
        if headers:
            base.update(headers)
        base.update(self.signed_headers(body))
        return base

    def wrap(self, upstream_url: str) -> str:
        """Wrap an upstream URL. Same as guard() but uses instance config."""
        return guard(upstream_url, self.proxy_url, self.agent_id)

    def list_escalations(self) -> list:
        """Fetch pending escalations from AgentGuard."""
        import urllib.request
        import json

        req = urllib.request.Request(
            f"{self.proxy_url}/escalations",
            headers={"X-AgentGuard-Admin-Key": self.admin_key},
        )
        try:
            with urllib.request.urlopen(req, timeout=5) as resp:
                data = json.loads(resp.read())
                return data.get("escalations", [])
        except Exception:
            return []

    def approve(self, escalation_id: str) -> bool:
        """Approve a pending escalation."""
        return self._resolve(escalation_id, "approve")

    def reject(self, escalation_id: str) -> bool:
        """Reject a pending escalation."""
        return self._resolve(escalation_id, "reject")

    def health(self) -> dict:
        """Check AgentGuard health."""
        import urllib.request
        import json
        try:
            with urllib.request.urlopen(
                f"{self.proxy_url}/health", timeout=3
            ) as resp:
                return json.loads(resp.read())
        except Exception as e:
            return {"status": "unreachable", "error": str(e)}

    def _resolve(self, escalation_id: str, action: str) -> bool:
        import urllib.request
        req = urllib.request.Request(
            f"{self.proxy_url}/escalations/{escalation_id}/{action}",
            data=b"",
            method="POST",
            headers={"X-AgentGuard-Admin-Key": self.admin_key},
        )
        try:
            with urllib.request.urlopen(req, timeout=5) as resp:
                return resp.status == 200
        except Exception:
            return False
